lib/rucio/core/config.py

```diff
@@ -423,7 +429,32 @@ def remove_option(section: str, option: str, *, session: "Session") -> bool:
         return True
 
 
-def read_from_cache(key: str, expiration_time: int = 900) -> Any:
+@overload
+def read_from_cache(key: Literal['sections'], expiration_time: int = 900) -> Union[list[str], NoValue]:
+    ...
+
+
+@overload
+def read_from_cache(key: OptionsKey, expiration_time: int = 900) -> Union[list[str], NoValue]:
+    ...
+
+
+@overload
+def read_from_cache(key: HasKey, expiration_time: int = 900) -> Union[bool, NoValue]:
+    ...
+
+
+@overload
+def read_from_cache(key: ItemsKey, expiration_time: int = 900) -> Union[Sequence[Row], NoValue]:
+    ...
+
+
+@overload
+def read_from_cache(key: ValueKey, expiration_time: int = 900) -> Union[Any, NoValue]:
```

Comment:
Because all of these return types are `Union[..., NoValue]`, you could create a `TypeVar` to parametrize this type, see https://stackoverflow.com/a/46384213

e.g. something like

```python
T = TypeVar('T')

CachedValue = Union[T, NoValue]

...

def read_from_cache(key: HasKey, expiration_time: int = 900) -> CachedValue[bool]:

...
```

Reply:
Something also worth mentioning is that the docstring for `NoValue` says:
```
    """Describe a missing cache value.

    The :data:`.NO_VALUE` constant should be used.

    """
```

So it might be better to use something like `Literal[NO_VALUE]`? Not completely sure about this, though. We use `NoValue` quite a bit at the moment, so you can probably ignore this for now and at a later point we could modify all `NoValue` annotations to `Literal[NO_VALUE]` if needed